native_client/deepspeech.cc

```diff
@@ -21,347 +23,670 @@
   }
 
-  char* output = (char*)malloc(sizeof(char) * decoded_length);
-  char* pen = output;
-  for (int i = 0; i < output_length - 1; i++) {
-    int64 character = decoder_outputs[0][0][i];
-    const std::string& str = mPriv->alphabet->StringFromLabel(character);
-    strncpy(pen, str.c_str(), str.size());
-    pen += str.size();
+  return strdup(output.str().c_str());
+}
+
+int
+DS_CreateModel(char* aModelPath,
+               int aNCep,
+               int aNContext,
+               char* aAlphabetConfigPath,
+               int aBeamWidth,
+               ModelState** retval)
+{
+  ModelState* model = new ModelState();
+  model->mmap_env   = new MemmappedEnv(Env::Default());
+  model->ncep       = aNCep;
+  model->ncontext   = aNContext;
+  model->alphabet   = new Alphabet(aAlphabetConfigPath);
+  model->beam_width = aBeamWidth;
+  model->run_aot    = false;
+
+  *retval = nullptr;
+
+  DS_PrintVersions();
+
+  if (!aModelPath || strlen(aModelPath) < 1) {
+    std::cerr << "No model specified, will rely on built-in model." << std::endl;
+    model->run_aot = true;
+    return 0;
+  }
+
+  Status status;
+  SessionOptions options;
+
+  bool is_mmap = std::string(aModelPath).find(".pbmm") != std::string::npos;
+  if (!is_mmap) {
+    std::cerr << "Warning: reading entire model file into memory. Transform model file into an mmapped graph to reduce heap usage." << std::endl;
+  } else {
+    status = model->mmap_env->InitializeFromFile(aModelPath);
+    if (!status.ok()) {
+      std::cerr << status << std::endl;
+      return status.code();
+    }
+
+    options.config.mutable_graph_options()
+      ->mutable_optimizer_options()
+      ->set_opt_level(::OptimizerOptions::L0);
+    options.env = model->mmap_env;
+  }
+
+  status = NewSession(options, &model->session);
+  if (!status.ok()) {
+    std::cerr << status << std::endl;
+    return status.code();
+  }
+
+  if (is_mmap) {
+    status = ReadBinaryProto(model->mmap_env,
+                             MemmappedFileSystem::kMemmappedPackageDefaultGraphDef,
+                             &model->graph_def);
+  } else {
+    status = ReadBinaryProto(Env::Default(), aModelPath, &model->graph_def);
+  }
+  if (!status.ok()) {
+    std::cerr << status << std::endl;
+    delete model;
+    return status.code();
   }
-  *pen = '\0';
 
-  for (int i = 0; i < timesteps; ++i) {
-    for (int j = 0; j < batch_size; ++j) {
-      free(aLogits[i][j]);
+  status = model->session->Create(model->graph_def);
+  if (!status.ok()) {
+    std::cerr << status << std::endl;
+    delete model;
+    return status.code();
+  }
+
+  for (int i = 0; i < model->graph_def.node_size(); ++i) {
+    NodeDef node = model->graph_def.node(i);
+    if (node.name() == "input_node") {
+      const auto& shape = node.attr().at("shape").shape();
+      model->n_steps = shape.dim(1).size();
+      model->mfcc_feats_per_timestep = shape.dim(2).size();
+      // mfcc_features_per_timestep = MFCC_FEATURES * ((2*n_context) + 1)
```

Comment:
leftover comment ?

Reply:
No, it's explaining the math on the line below.